File: impls/objpascal/step9_try.pas
program Mal;

{$H+} // Use AnsiString

Uses sysutils,
     CMem,
     fgl,
     math,
     mal_readline,
     mal_types,
     mal_func,
     reader,
     printer,
     mal_env,
     core;

var
    Repl_Env : TEnv;
    Line     : string;
    I        : longint;
    Key      : string;
    CmdArgs  : TMalArray;

// read
function READ(const Str: string) : TMal;
begin
    READ := read_str(Str);
end;

// eval

function starts_with(Ast: TMal; Sym: String) : Boolean;
var
   Arr : TMalArray;
   A0  : TMal;
begin
   if Ast.ClassType <> TMalList then Exit (False);
   Arr := (Ast as TMalList).Val;
   if Length (Arr) = 0 then Exit (False);
   A0 := Arr [0];
   starts_with := (A0.ClassType = TMalSymbol) and ((A0 as TMalSymbol).Val = Sym);
end;

function quasiquote(Ast: TMal) : TMal;
var
    Arr      : TMalArray;
    Res, Elt : TMal;
    I        : longint;
begin
    if Ast is TMalSymbol or Ast is TMalHashMap then
        Exit(_list(TMalSymbol.Create('quote'), Ast));

    if not (Ast is TMalList) then
        Exit(Ast);

    Arr := (Ast as TMalList).Val;
    if starts_with (Ast, 'unquote') then Exit(Arr[1]);

    Res := _list();
    for I := 1 to Length(Arr) do
    begin
        Elt := Arr [Length(Arr) - I];
        if starts_with (Elt, 'splice-unquote') then
            Res := _list(TMalSymbol.Create('concat'), (Elt as TMalList).Val[1], Res)
        else
            Res := _list(TMalSymbol.Create('cons'), quasiquote (Elt), Res);
    end;
    if Ast.ClassType <> TMalList then
        Exit(_list(TMalSymbol.Create('vec'), Res))
    else
        Exit(Res);
end;

function is_macro_call(Ast: TMal; Env: TEnv): Boolean;
var
    A0  : TMal;
    Mac : TMal;
begin
    is_macro_call := false;
    if (Ast.ClassType = TMalList) and
       (Length((Ast as TMalList).Val) > 0) then
    begin
        A0 := (Ast as TMalList).Val[0];
        if (A0 is TMalSymbol) and
           (Env.Find(A0 as TMalSymbol) <> nil) then
        begin
            Mac := Env.Get((A0 as TMalSymbol));
            if Mac is TMalFunc then
                is_macro_call := (Mac as TMalFunc).isMacro;
        end;
    end;

end;

// Forward declation since eval_ast call it
function EVAL(Ast: TMal; Env: TEnv) : TMal; forward;

function macroexpand(Ast: TMal; Env: TEnv): TMal;
var
    A0   : TMal;
    Arr  : TMalArray;
    Args : TMalArray;
    Mac  : TMalFunc;
begin
    while is_macro_call(Ast, Env) do
    begin
        Arr := (Ast as TMalList).Val;
        A0 := Arr[0];
        Mac := Env.Get((A0 as TMalSymbol)) as TMalFunc;
        Args := (Ast as TMalList).Rest.Val;
        if Mac.Ast = nil then
            Ast := Mac.Val(Args)
        else
            Ast := EVAL(Mac.Ast,
                        TEnv.Create(Mac.Env, Mac.Params, Args));
    end;
    macroexpand := Ast;
end;

function eval_ast(Ast: TMal; Env: TEnv) : TMal;
var
    OldArr, NewArr   : TMalArray;
    OldDict, NewDict : TMalDict;
    I                : longint;
begin
    if Ast is TMalSymbol then
    begin
        eval_ast := Env.Get((Ast as TMalSymbol));
    end
    else if Ast is TMalList then
    begin
        OldArr := (Ast as TMalList).Val;
        SetLength(NewArr, Length(OldArr));
        for I := 0 to Length(OldArr)-1 do
        begin
            NewArr[I] := EVAL(OldArr[I], Env);
        end;
        if Ast is TMalVector then
            eval_ast := TMalVector.Create(NewArr)
        else
            eval_ast := TMalList.Create(NewArr);
    end
    else if Ast is TMalHashMap then
    begin
        OldDict := (Ast as TMalHashMap).Val;
        NewDict := TMalDict.Create;
        I := 0;
        while I < OldDict.Count do
        begin
            NewDict[OldDict.Keys[I]] := EVAL(OldDict[OldDict.Keys[I]], Env);
            I := I + 1;
        end;
        eval_ast := TMalHashMap.Create(NewDict);
    end
    else
        eval_ast := Ast;
end;

function EVAL(Ast: TMal; Env: TEnv) : TMal;
var
    Lst    : TMalList;
    Arr    : TMalArray;
    Arr1   : TMalArray;
    A0Sym  : string;
    LetEnv : TEnv;
    Cond   : TMal;
    I      : longint;
    Fn     : TMalFunc;
    Args   : TMalArray;
    Err    : TMalArray;
begin
  while true do
  begin
    if Ast.ClassType <> TMalList then
        Exit(eval_ast(Ast, Env));

    Ast := macroexpand(Ast, Env);
    if Ast.ClassType <> TMalList then
        Exit(eval_ast(Ast, Env));

    // Apply list
    Lst := (Ast as TMalList);
    Arr := Lst.Val;
    if Length(Arr) = 0 then
        Exit(Ast);
    if Arr[0] is TMalSymbol then
        A0Sym := (Arr[0] as TMalSymbol).Val
    else
        A0Sym := '__<*fn*>__';

    case A0Sym of
    'def!':
        Exit(Env.Add((Arr[1] as TMalSymbol), EVAL(Arr[2], ENV)));
    'let*':
        begin
            LetEnv := TEnv.Create(Env);
            Arr1 := (Arr[1] as TMalList).Val;
            I := 0;
            while I < Length(Arr1) do
            begin
                LetEnv.Add((Arr1[I] as TMalSymbol), EVAL(Arr1[I+1], LetEnv));
                Inc(I,2);
            end;
            Env := LetEnv;
            Ast := Arr[2]; // TCO
        end;
    'quote':
        Exit(Arr[1]);
    'quasiquoteexpand':
        Exit(quasiquote(Arr[1]));
    'quasiquote':
        Ast := quasiquote(Arr[1]);
    'defmacro!':
    begin
        Fn := EVAL(Arr[2], ENV) as TMalFunc;
        Fn := TMalFunc.Clone(Fn);
        Fn.isMacro := true;
        Exit(Env.Add((Arr[1] as TMalSymbol), Fn));
    end;
    'macroexpand':
        Exit(macroexpand(Arr[1], Env));
    'try*':
    begin
        try
            Exit(EVAL(Arr[1], Env));
        except
            On E : Exception do
            begin
                if Length(Arr) < 3 then
                    raise;
                SetLength(Err, 1);
                if E.ClassType = TMalException then
                    Err[0] := (E as TMalException).Val
                else
                    Err[0] := TMalString.Create(E.message);
                Arr := (Arr[2] as TMalList).Val;
                Exit(EVAL(Arr[2], TEnv.Create(Env,
                                              _list(Arr[1]),
                                              Err)));
            end;
        end;
    end;
    'do':
        begin
            eval_ast(TMalList.Create(copy(Arr,1, Length(Arr)-2)), Env);
            Ast := Arr[Length(Arr)-1]; // TCO
        end;
    'if':
        begin
            Cond := EVAL(Arr[1], Env);
            if (Cond is TMalNil) or (Cond is TMalFalse) then
                if Length(Arr) > 3 then
                    Ast := Arr[3] // TCO
                else
                    Exit(TMalNil.Create)
            else
                Ast := Arr[2]; // TCO
        end;
    'fn*':
        begin
            Exit(TMalFunc.Create(Arr[2], Env, (Arr[1] as TMalList)));
        end;
    else
        begin
            Arr := (eval_ast(Ast, Env) as TMalList).Val;
            if Arr[0] is TMalFunc then
            begin
                Fn := Arr[0] as TMalFunc;
                if Length(Arr) < 2 then
                    SetLength(Args, 0)
                else
                    Args := copy(Arr, 1, Length(Arr)-1);
                if Fn.Ast = nil then
                    Exit(Fn.Val(Args))
                else
                begin
                    Env := TEnv.Create(Fn.Env, Fn.Params, Args);
                    Ast := Fn.Ast; // TCO
                end

            end
            else
                raise Exception.Create('invalid apply');
        end;
    end;
  end;
end;

// print
function PRINT(Exp: TMal) : string;
begin
    PRINT := pr_str(Exp, True);
end;

// repl
function REP(Str: string) : string;
begin
    REP := PRINT(EVAL(READ(Str), Repl_Env));
end;

function do_eval(Args : TMalArray) : TMal;
begin
    do_eval := EVAL(Args[0], Repl_Env);
end;

begin
    Repl_Env := TEnv.Create;
    core.EVAL := @EVAL;

    // core.pas: defined using Pascal
    for I := 0 to core.NS.Count-1 do
    begin
        Key := core.NS.Keys[I];
        Repl_Env.Add(TMalSymbol.Create(Key),
                     TMalFunc.Create(core.NS[Key]));
    end;
    Repl_Env.Add(TMalSymbol.Create('eval'), TMalFunc.Create(@do_eval));
    SetLength(CmdArgs, Max(0, ParamCount-1));
    for I := 2 to ParamCount do
        CmdArgs[I-2] := TMalString.Create(ParamStr(I));
    Repl_Env.Add(TMalSymbol.Create('*ARGV*'), TMalList.Create(CmdArgs));

    // core.mal: defined using language itself
    REP('(def! not (fn* (a) (if a false true)))');
    REP('(def! load-file (fn* (f) (eval (read-string (str "(do " (slurp f) "\nnil)")))))');
    REP('(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list ''if (first xs) (if (> (count xs) 1) (nth xs 1) (throw "odd number of forms to cond")) (cons ''cond (rest (rest xs)))))))');


    if ParamCount >= 1 then
    begin
        REP('(load-file "' + ParamStr(1) + '")');
        ExitCode := 0;
        Exit;
    end;

    while True do
    begin
        try
            Line := _readline('user> ');
            if Line = '' then continue;
            WriteLn(REP(Line))
        except
            On E : MalEOF do Halt(0);
            On E : Exception do
            begin
                if E.ClassType = TMalException then
                    WriteLn('Error: ' + pr_str((E as TMalException).Val, True))
                else
                    WriteLn('Error: ' + E.message);
                WriteLn('Backtrace:');
                WriteLn(GetBacktrace(E));
            end;
        end;
    end;
end.

File: impls/julia/step9_try.jl
#!/usr/bin/env julia

push!(LOAD_PATH, pwd(), "/usr/share/julia/base")
import readline_mod
import reader
import printer
using env
import core
using types

# READ
function READ(str)
    reader.read_str(str)
end

# EVAL
function quasiquote_loop(elts)
    acc = Any[]
    for i in length(elts):-1:1
        elt = elts[i]
        if isa(elt, Array) && length(elt) == 2 && elt[1] == symbol("splice-unquote")
            acc = Any[:concat, elt[2], acc]
        else
            acc = Any[:cons, quasiquote(elt), acc]
        end
    end
    return acc
end

function quasiquote(ast)
    if isa(ast, Array)
        if length(ast) == 2 && ast[1] == symbol("unquote")
            ast[2]
        else
            quasiquote_loop(ast)
        end
    elseif isa(ast, Tuple)
        Any[:vec, quasiquote_loop(ast)]
    elseif typeof(ast) == Symbol || isa(ast, Dict)
        Any[:quote, ast]
    else
        ast
    end
end

function ismacroCall(ast, env)
    return isa(ast, Array) &&
           !isempty(ast) &&
           isa(ast[1], Symbol) &&
           env_find(env, ast[1]) != nothing &&
           isa(env_get(env, ast[1]), MalFunc) &&
           env_get(env, ast[1]).ismacro
end

function macroexpand(ast, env)
    while ismacroCall(ast, env)
        mac = env_get(env, ast[1])
        ast = mac.fn(ast[2:end]...)
    end
    ast
end

function eval_ast(ast, env)
    if typeof(ast) == Symbol
        env_get(env,ast)
    elseif isa(ast, Array) || isa(ast, Tuple)
        map((x) -> EVAL(x,env), ast)
    elseif isa(ast, Dict)
        [x[1] => EVAL(x[2], env) for x=ast]
    else
        ast
    end
end

function EVAL(ast, env)
  while true
    #println("EVAL: $(printer.pr_str(ast,true))")
    if !isa(ast, Array) return eval_ast(ast, env) end

    # apply
    ast = macroexpand(ast, env)
    if !isa(ast, Array) return eval_ast(ast, env) end
    if isempty(ast) return ast end

    if     :def! == ast[1]
        return env_set(env, ast[2], EVAL(ast[3], env))
    elseif symbol("let*") == ast[1]
        let_env = Env(env)
        for i = 1:2:length(ast[2])
            env_set(let_env, ast[2][i], EVAL(ast[2][i+1], let_env))
        end
        env = let_env
        ast = ast[3]
        # TCO loop
    elseif :quote == ast[1]
        return ast[2]
    elseif :quasiquoteexpand == ast[1]
        return quasiquote(ast[2])
    elseif :quasiquote == ast[1]
        ast = quasiquote(ast[2])
        # TCO loop
    elseif :defmacro! == ast[1]
        func = EVAL(ast[3], env)
        func.ismacro = true
        return env_set(env, ast[2], func)
    elseif :macroexpand == ast[1]
        return macroexpand(ast[2], env)
    elseif symbol("try*") == ast[1]
        try
            return EVAL(ast[2], env)
        catch exc
            e = string(exc)
            if isa(exc, MalException)
                e = exc.malval
            elseif isa(exc, ErrorException)
                e = exc.msg
            else
                e = string(e)
            end
            if length(ast) > 2 && ast[3][1] == symbol("catch*")
                return EVAL(ast[3][3], Env(env, Any[ast[3][2]], Any[e]))
            else
                rethrow(exc)
            end
        end
    elseif :do == ast[1]
        eval_ast(ast[2:end-1], env)
        ast = ast[end]
        # TCO loop
    elseif :if == ast[1]
        cond = EVAL(ast[2], env)
        if cond === nothing || cond === false
            if length(ast) >= 4
                ast = ast[4]
                # TCO loop
            else
                return nothing
            end
        else
            ast = ast[3]
            # TCO loop
        end
    elseif symbol("fn*") == ast[1]
        return MalFunc(
            (args...) -> EVAL(ast[3], Env(env, ast[2], Any[args...])),
            ast[3], env, ast[2])
    else
        el = eval_ast(ast, env)
        f, args = el[1], el[2:end]
        if isa(f, MalFunc)
            ast = f.ast
            env = Env(f.env, f.params, args)
            # TCO loop
        else
            return f(args...)
        end
    end
  end
end

# PRINT
function PRINT(exp)
    printer.pr_str(exp)
end

# REPL
repl_env = nothing
function REP(str)
    return PRINT(EVAL(READ(str), repl_env))
end

# core.jl: defined using Julia
repl_env = Env(nothing, core.ns)
env_set(repl_env, :eval, (ast) -> EVAL(ast, repl_env))
env_set(repl_env, symbol("*ARGV*"), ARGS[2:end])

# core.mal: defined using the language itself
REP("(def! not (fn* (a) (if a false true)))")
REP("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))")
REP("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))")


if length(ARGS) > 0
    REP("(load-file \"$(ARGS[1])\")")
    exit(0)
end

while true
    line = readline_mod.do_readline("user> ")
    if line === nothing break end
    try
        println(REP(line))
    catch e
        if isa(e, ErrorException)
            println("Error: $(e.msg)")
        else
            println("Error: $(string(e))")
        end
        # TODO: show at least part of stack
        if !isa(e, StackOverflowError)
            bt = catch_backtrace()
            Base.show_backtrace(STDERR, bt)
        end
        println()
    end
end
